### PR TITLE
Redesign Code Detection System for Improved Accuracy and Performance

### PR DESCRIPTION
**Pull request by Arborym**
Discussion here: https://discord.com/channels/238956364729155585/1040360228388098159/1138249299847827498

Key insights:
- Use distinct detection algorithms for each of the server's most popular languages (Python, Java, C/C++, JavaScript, HTML, and CSS). This will improve the detection system's accuracy.
- Begin by not deleting messages for a month (or longer) to ensure that it works 100% of the time. This will help ensure the new system's dependability.
- Create our own code detection algorithms. This is more precise than using third-party libraries.
- For decision-making, use simple statistics. This will aid in lowering the number of false positives.

Sometimes, people ask questions & don't format their code:
image

This PR implements a system to automatically detect and format people's code for them:
image

### CODE REVIEW

**Comment by GabeMillikan** [comment]
Quick preview for channel permissions:

https://github.com/Arborym/Pax-Academia/assets/44247924/82b18237-3d08-4989-92b8-a26421a79b3f

```py
UNFORMATTED_CODE_DETECTION_CATEGORY_ID=<computer science help category id>
AUTO_FORMAT_CODE_CHANNEL_IDS=<python channel id>
```

**Comment by sebastiaan-daniels** [comment]
I have a feeling this would best be ephemeral. If people send very long messages the bot would respond with a message just as long, which might clutter the chat.

**Comment by GabeMillikan** [comment]
> I have a feeling this would best be ephemeral. If people send very long messages the bot would respond with a message just as long, which might clutter the chat.

I think you're right about it probably cluttering chat, but I'm tempted to think that people _still_ won't bother formatting their code. This message actually serves a purpose in allowing helpers to actually read the code as originally written.

Also, I don't think it's possible to send an ephemeral message if the user doesn't use a slash command

**Comment by sebastiaan-daniels** [comment]
> > I have a feeling this would best be ephemeral. If people send very long messages the bot would respond with a message just as long, which might clutter the chat.
> 
> I think you're right about it probably cluttering chat, but I'm tempted to think that people _still_ won't bother formatting their code. This message actually serves a purpose in allowing helpers to actually read the code as originally written.
> 
> Also, I don't think it's possible to send an ephemeral message if the user doesn't use a slash command

Alrighty! We can definitely implement it like this then and change it in case of complaints. I thought it was possible to respond ephemerally even in on_message. I think I saw it before with some bot?

Also please undo the renaming of staffapps. Front-end and backend are technical terms. Both files are technically both front and backend. You could rename it to staffapps-user and staffapps-staff instead or keep the original.

**Comment by sebastiaan-daniels** [comment]
Not me completely messing up in githubs GUI :monkas:

File: cogs/staffapps_backend.py
Comment (sebastiaan-daniels) [inline cogs/staffapps_backend.py:1]: Backoffice != backend

File: cogs/staffapps_frontend.py
Comment (sebastiaan-daniels) [inline cogs/staffapps_frontend.py:1]: StaffAppsUser != frontend
https://www.computerscience.org/bootcamps/resources/frontend-vs-backend/#:~:text=What's%20the%20Difference%20Between%20Front,system%2C%20data%2C%20and%20logic.

**Review by sebastiaan-daniels — APPROVED** [review]

**Pull request merged**